parsec/scheduling.c

```diff
@@ -532,6 +531,108 @@ int __parsec_task_progress( parsec_execution_stream_t* es,
     return rc;
 }
 
+static int __parsec_taskpool_test( parsec_taskpool_t* tp, parsec_execution_stream_t *es )
+{
+    parsec_context_t* parsec_context = es->virtual_process->parsec_context;
+    parsec_task_t* task;
+    int nbiterations = 0, distance, rc;
+
+    assert(PARSEC_THREAD_IS_MASTER(es));
+
+    /* first select begin, right before the wait_for_the... goto label */
+    PARSEC_PINS(es, SELECT_BEGIN, NULL);
+
+    if( NULL == parsec_current_scheduler ) {
+        parsec_fatal("Main thread entered parsec_taskpool_test, while a scheduler is not selected yet!");
+        return -1;
+    }
+
+    if( tp->tdm.module->taskpool_state(tp) != PARSEC_TERM_TP_TERMINATED ) {
+        /* Here we detach all dtd taskpools registered with us */
+        parsec_detach_all_dtd_taskpool_from_context(parsec_context);
+
+#if defined(DISTRIBUTED)
+        if( (1 == parsec_communication_engine_up) &&
+            (es->virtual_process[0].parsec_context->nb_nodes == 1)  ) {
+            /* check for remote deps completion */
+            while(parsec_remote_dep_progress(es) > 0) /* nothing */;
+        }
+#endif /* defined(DISTRIBUTED) */
+
+        task = parsec_current_scheduler->module.select(es, &distance);
+
+        if( task != NULL ) {
+            rc = __parsec_task_progress(es, task, distance);
+            (void)rc;  /* for now ignore the return value */
+
+            nbiterations++;
+        }
+    }
+
+    PARSEC_PINS(es, SELECT_END, NULL);
+
+    return nbiterations;
+}
+
+static int __parsec_taskpool_wait( parsec_taskpool_t* tp, parsec_execution_stream_t *es )
+{
+    uint64_t misses_in_a_row;
+    parsec_task_t* task;
+    int nbiterations = 0, distance, rc;
+    struct timespec rqtp;
+
+    rqtp.tv_sec = 0;
+    misses_in_a_row = 1;
+
+    assert(PARSEC_THREAD_IS_MASTER(es));
```

Comment:
don't think that assert is correct, what about recursive task pools

Reply:
Unlike in test I don't think we should allow any other threads to do a blocking operation on any parsec_objects. We will run out of working threads, and encounter undesired side-effects.